query-planner-js/src/buildPlan.ts

```diff
@@ -177,7 +177,7 @@ class QueryPlanningTaversal<RV extends Vertex> {
         // Do note that we'll only need that `__typename` if there is no other selections inside `foo`, and so we might include
         // it unecessarally in practice: it's a very minor inefficiency though.
         if (operation.kind === 'FragmentElement') {
-          this.closedBranches.push([option.paths.map(p => terminateWithNonRequestedTypenameField(p))]);
+          this.closedBranches.push(options.map((o) => o.paths.map(p => terminateWithNonRequestedTypenameField(p))));
```

Comment:
Remark: this change is technically not really necessary to this patch, and in fact I'm not entirely sure if this whole branch of query planner can still be hit with the validation added (but I'm unsure and would rather keep it for now), but I noticed this while working on this patch and the new version is "more correct" (mostly, it avoids producing plans that are more inefficient than necessary if this code path is triggered).